### PR TITLE
[FEAT][BWS]: fcm data messages

### PR DESCRIPTION
**Pull request by rastajpa**

### CODE REVIEW

File: packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts [hunk 1]
```diff
@@ -168,27 +208,33 @@ export class PushNotificationsService {
...
-                        },
                         data: {
-                          walletId: sjcl.codec.hex.fromBits(sjcl.hash.sha256.hash(notification.walletId)),
+                          walletId: sjcl.codec.hex.fromBits(sjcl.hash.sha256.hash(recipient.walletId)),
                           tokenAddress,
                           multisigContractAddress,
                           copayerId: sjcl.codec.hex.fromBits(sjcl.hash.sha256.hash(recipient.copayerId)),
-                          title: content.plain.subject,
-                          body: content.plain.body,
+                          title: content?.plain?.subject,
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts:221]: ❤️

File: packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts [hunk 2]
```diff
@@ -231,64 +277,96 @@ export class PushNotificationsService {
...
-        unit = wallet.coin;
-      }
+    if (notification.type === 'NewBlock') {
+      this.storage.fetchLatestPushNotificationSubs((err, subs) => {
+        if (err) return cb(err);
 
-      this.storage.fetchPreferences(notification.walletId, null, (err, preferences) => {
-        if (err) logger.error(err);
-        if (_.isEmpty(preferences)) preferences = [];
+        async.map(
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts:291]: Add a log.info before this: "Sending new block [coin/chain] notifications to: xxx subscribers"

**Review by gonzaloleon — APPROVED** [review]
ACK! Nice feature!

**Review by gabrielbazan7 — APPROVED** [review]
ACK !

File: packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts [hunk 3]
```diff
@@ -231,64 +277,99 @@ export class PushNotificationsService {
   }
 
   _checkShouldSendNotif(notification, cb) {
-    if (notification.type != 'NewTxProposal') return cb(null, true);
+    if (notification.type != 'NewTxProposal' || notification.type == 'NewBlock') return cb(null, true);
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts:280]: remove ||

**Comment by SatoshiNakamotogmx** [comment]
Remove

On Dec 21, 2020 10:09 AM, "ematiu" <notifications@github.com> wrote:

*@matiu* commented on this pull request.
------------------------------

In packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts
<https://github.com/bitpay/bitcore/pull/3012#discussion_r546853059>:

> @@ -231,64 +277,99 @@ export class PushNotificationsService {
   }

   _checkShouldSendNotif(notification, cb) {
-    if (notification.type != 'NewTxProposal') return cb(null, true);
+    if (notification.type != 'NewTxProposal' || notification.type ==
'NewBlock') return cb(null, true);

remove ||

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/bitpay/bitcore/pull/3012#pullrequestreview-556569091>,
or unsubscribe

**Comment by SatoshiNakamotogmx** [comment]
Thank you very much

File: packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts [hunk 4]
```diff
@@ -150,54 +188,49 @@ export class PushNotificationsService {
...
-                        }
-                      };
-                    });
-                    return next(err, notifications);
-                  });
-                },
-                (err, allNotifications) => {
-                  if (err) return next(err);
-                  return next(null, _.flatten(allNotifications));
+              this._getSubscriptions(notification, recipientsList, contents, next);
```
Comment (matiu) [inline packages/bitcore-wallet-service/src/lib/pushnotificationsservice.ts:196]: nice!

File: packages/bitcore-wallet-service/test/integration/pushNotifications.js [hunk 5]
```diff
@@ -742,12 +862,113 @@ describe('Push notifications', function() {
...
+            }, function(err) {
+              should.not.exist(err);
+              done();
+            });
+          });
+        });
+      });
+    });
+    
+    it('should notify NewBlock to all devices subscribed in the last 10 minutes', function(done) {
```
Comment (matiu) [inline packages/bitcore-wallet-service/test/integration/pushNotifications.js:936]: nice!

**Pull request merged**